src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -485,14 +485,20 @@ public void SetStateMachine(IAsyncStateMachine stateMachine)
             // object's identity to track this specific builder/state machine.  As such, we proceed to
             // overwrite whatever's there anyway, even if it's non-null.
             var box = AsyncMethodBuilderCore.TrackAsyncMethodCompletion ?
-                new DebugFinalizableAsyncStateMachineBox<TStateMachine>() :
+                CreateDebugFinalizableAsyncStateMachineBox<TStateMachine>() :
                 new AsyncStateMachineBox<TStateMachine>();
             m_task = box; // important: this must be done before storing stateMachine into box.StateMachine!
             box.StateMachine = stateMachine;
             box.Context = currentContext;
             return box;
         }
 
+        // Avoid forcing the JIT to build DebugFinalizableAsyncStateMachineBox<TStateMachine> unless it's actually needed.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static AsyncStateMachineBox<TStateMachine> CreateDebugFinalizableAsyncStateMachineBox<TStateMachine>()
```

Comment:
Call this method in `InitializeTaskAsStateMachineBox` as well?

Reply:
I guess it doesn't hurt.  That method is only invoked while debugging, but yeah, I've changed it.